keras/src/layers/rnn/rnn.py

```diff
@@ -382,6 +382,25 @@ def call(
                 initial_state = self.get_initial_state(
                     batch_size=ops.shape(sequences)[0]
                 )
+        if self.stateful:
+            actual_batch_size = ops.shape(sequences)[0]
+            expected_batch_size = (
```

Comment:
You should compute this when we create the states, and cache it as a property (it's expensive to recompute at every call in eager mode)

Reply:
Thank you! Just updated by adding a property `_expected_batch_size` which computes this in the `build` method after `states` is set.